XMonad/Hooks/BorderPerWindow.hs

```diff
@@ -0,0 +1,132 @@
+
+                                      -- * Why Not Something Simpler?
+                                      -- $options
+                                      defineWindowBorder
+                                    , setBorderPerWindow
+                                    ) where
+
+import XMonad
+import XMonad.Layout.LayoutModifier
+import Graphics.X11.Xlib (Dimension)
+
+-- $usage
+--
+-- 1. Specify which window to customize border in 'manageHook' of the config:
+--
+-- > myManageHook :: ManageHook
+-- > myManageHook = composeAll
+-- >     [ className =? "firefox"  --> defineWindowBorder 0
+-- >     , className =? "Chromium" --> defineWindowBorder 0
+-- >     , isDialog                --> defineWindowBorder 8
+-- >     ]
+--
+-- 2. Execute the border setter at layout handling hook:
+--
+-- > myLayoutHook = setBorderPerWindow
+-- >              $ Tall 1 (3/100) (1/2) ||| Full ||| Grid ||| ...
+--
+-- 3. Add the two hooks in your config:
+--
+-- > myConfig = def
+-- >     { ...
+-- >     , layoutHook = myLayoutHook
+-- >     , manageHook = myManageHook
+-- >     , ...
+-- >     }
+
+-- $design
+--
+-- 1. Keep it simple. Since the extension does not aim to change border setting
+--    when layout changes, only execute the border setting function once to
+--    avoid potential window flashing/jumping/scaling.
+--
+-- 2. The 'ManageHook' eDSL is a nice language for specifying windows. Let's
+--    build on top of it and use it to specify window to define border.
+
+-- $options
+--
+-- 1. Just setting window-specific border in 'manageHook' does not work, since
+--    XMonad sets window border after 'ManageHook's are applied.
+--
+-- 2. 'XMonad.Layout.NoBorders' provides a way to set zero-width window border
+--    not permanently, e.g., need to recover border when layout changes, which
+--    results in complexity, e.g., need to set border every time a layout is
+--    rendered, and has a window jumping effect when switching between
+--    workspaces: for a zero-width border window, the border (global config) is
+--    temporarily drawn and then removed; all pixels of the window is scaled
+--    during the add-n-remove procedure, which could be noticeable.
+--
+-- 3. 'XMonad.Layout.VoidBorders' allows one to choose layout but not window.
+
+-- | The core type that records border width per window
+data WindowBorderSpec
+    = WindowBorderSpec Window Dimension deriving (Read, Show)
+
+-- | The XMonad message that transfers 'WindowBorderSpec' from the window
+-- creation event to the layout handling function.
+-- Note that XMonad reads config and draws window border after 'ManageHook's
+-- are applied. It does not work if we set the border in a 'ManageHook'
+-- (hooking window creation). So we choose to send a message in
+-- 'defineWindowBorder' function to a later stage, e.g., layout handling, to
+-- set the window border.
+data WindowBorderSpecMessage
+    = WindowBorderSpecMessage WindowBorderSpec deriving (Read, Show)
```

Comment:
We can probably combine these into a single `WindowBorder[Spec]Message` since they are not used separately

Reply:
The reason I introduced `WindowBorderSpec` is for
```
data BorderPerWindow a
    = BorderPerWindow (Maybe WindowBorderSpec) deriving (Read, Show)
```
I think we need a data type for `Maybe` other than two arguments. And since I introduce `WindowBorderSpec`, I just use it in `WindowBorderSpecMessage` to make it simpler.

I think the only chance to get rid of it is to eliminate it in both, which goes to your second comment. Please see my thoughts there.

Reply:
I meant it in the sense that we could just define

``` haskell
  data WindowBorderMessage
      = WindowBorderMessage Window Dimension deriving (Read, Show)

  instance Message WindowBorderMessage
```

and be on our way.  The definition of `BorderPerWindow` wouldn't need to change at all but it would safe us from having to wrap/unwrap twice a few times

Reply:
I think I may get a good opportunity to better understand a little more on GHC. Could you explain a little more on wrap/unwarp twice? Or any article I can read? I really want to make it fast if possible, at least not adding too much overhead because of the module.

Reply:
Every time you create a new type with `data` you're essentially adding one layer of indirection (i.e., one more pointer to follow) to what you're handling.  Wrapping and unwrapping then _technically_ adds some costs because you need to follow the pointers to the actual object that you want to look at and doing that twice is more expensive than only doing it once :)

There is `newtype` for wrappers around a single field, which then has the same runtime representation as the thing that it wraps.  For example, you could just let `WindowBorderSpecMessage` be a newtype and that would be as "efficient" as what I'm suggesting.

However, I don't think overhead/performance is the primary goal here; I was thinking more along the lines of convenience.  I feel that in this case it simply reduces mental overhead (one less type that looks a lot like the others to worry about).  Plus, instead of something like

```haskell
liftX . broadcastMessage . WindowBorderSpecMessage $ WindowBorderSpec w bw
```

one gets to write

```haskell
liftX . broadcastMessage $ WindowBorderSpecMessage w bw
```

Reply:
I see. Thanks for the explanation!
So you suggest instead of
```Haskell
    pureMess (BorderPerWindow _) msg
        | Just (WindowBorderSpecMessage spec) <- fromMessage msg =
            Just . BorderPerWindow $ Just spec
```
I can revise it to
```Haskell
    pureMess (BorderPerWindow _) msg
        | Just (WindowBorderSpecMessage w bw) <- fromMessage msg =
            Just . BorderPerWindow . Just $ WindowBorderSpec w bw
```
When I wrote the old code. I guess `Just spec` will ease the compiler's job---the compiler directly passes the address of `spec` from `WindowBorderSpecMessage` to `BorderPerWindow` without checking what's inside. And until the `redoLayout` function the `WindowBorderSpec` data needs to be unpacked to expose the internal structure with guards:
```Haskell
    redoLayout (BorderPerWindow (Just (WindowBorderSpec w bw))) _ _ wrs = do
```
Is my guess correct?

Reply:
> I see. Thanks for the explanation! So you suggest instead of
>
> […]
>
> I can revise it to
>
> ```haskell
>     pureMess (BorderPerWindow _) msg
>         | Just (WindowBorderSpecMessage w bw) <- fromMessage msg =
>             Just . BorderPerWindow . Just $ WindowBorderSpec w bw
> ```

I believe it would just be

```haskell
    pureMess (BorderPerWindow _) msg
        | Just spec@WindowBorderSpecMessage{} <- fromMessage msg =
            Just . BorderPerWindow $ Just spec
```

But really, just rename `WindowBorderSpec` to `WindowBorderSpecMessage` (and delete the other type, of course) and follow the compiler—it won't disappoint!

> When I wrote the old code. I guess `Just spec` will ease the compiler's job---the compiler directly passes the address of `spec` from `WindowBorderSpecMessage` to `BorderPerWindow` without checking what's inside. And until the `redoLayout` function the `WindowBorderSpec` data needs to be unpacked to expose the internal structure with guards:
>
> […]
>
> Is my guess correct?

Sure, matching on only the outer constructor is probably cheaper because we don't have to actually look into the structure (this is also done by `spec@WindowBorderSpecMessage{}` as it only checks the outer shape of the data).  Though this again only conceptually matters—practically, the call to `setWindowBorderWidth` will take the vast majority of the time anyways

Reply:
Oh, I see. Thanks, @slotThe! Will do.